api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java

```diff
@@ -165,35 +169,57 @@ public void close() {
     }
   }
 
-  private void scheduleTokensRenewal(Tokens currentTokens) {
-    Instant now = Instant.now();
+  private void maybeScheduleTokensRenewal(Tokens currentTokens) {
+    Instant now = clock.get();
+    if (idle(now)) {
+      sleeping.set(true);
```

Comment:
Probably unlikely, but I suspect this update can race with the conditional update in `authenticate()`. A user (indirectly) calls `authenticate()` - that thread gets stuck after the CAS - this function runs and sets sleeping back to true - and then the user thread resumes. It feels, that it can wake-up twice.

Reply:
> this function runs

In which thread? I don't think that's possible: if `sleeping` was set to true, this means that the background thread pool is idle, and there is no refresh task scheduled for later execution.

It could happen though, that a user thread enters `authenticate()`, resets `sleeping` to false, then renew the tokens (which involves network IO), then gets stuck for several seconds somewhere, then enters `maybeScheduleTokensRenewal`, then resets `sleeping` to true again. But that's not a pathological code path.